src/evo/cbtx.cpp

```diff
@@ -331,13 +331,33 @@ bool CheckCbTxBestChainlock(const CBlock& block, const CBlockIndex* pindex,
         return true;
     }
 
+    static Mutex cached_mutex;
+    static const CBlockIndex* cached_pindex GUARDED_BY(cached_mutex){nullptr};
+    static std::optional<std::pair<CBLSSignature, uint32_t>> cached_chainlock GUARDED_BY(cached_mutex){std::nullopt};
+
     auto best_clsig = chainlock_handler.GetBestChainLock();
     if (best_clsig.getHeight() == pindex->nHeight - 1 && cbTx.bestCLHeightDiff == 0 && cbTx.bestCLSignature == best_clsig.getSig()) {
         // matches our best clsig which still hold values for the previous block
+        {
+            LOCK(cached_mutex);
+            cached_chainlock = std::make_pair(cbTx.bestCLSignature, cbTx.bestCLHeightDiff);
+            cached_pindex = pindex;
+        }
         return true;
     }
 
-    const auto prevBlockCoinbaseChainlock = GetNonNullCoinbaseChainlock(pindex->pprev);
+    std::optional<std::pair<CBLSSignature, uint32_t>> prevBlockCoinbaseChainlock{std::nullopt};
+    bool use_cached = false;
+    {
+        LOCK(cached_mutex);
+        if (cached_pindex == pindex->pprev) {
+            use_cached = true;
+            prevBlockCoinbaseChainlock = cached_chainlock;
+        }
+    }
```

Comment:
or something like that 
```suggestion
auto [prevBlockCoinbaseChainlock, use_cached] = [&]() {
    LOCK(cached_mutex);
    if (cached_pindex == pindex->pprev) {
        return std::make_pair(cached_chainlock, true);
    }
    return std::make_pair(std::nullopt, false);
}();```